factorio.tf

```
@@ -82,68 +58,83 @@ resource "aws_iam_role_policy_attachment" "factorio" {
+  state_device_name = "/dev/sdf"
+  ami_device = [
+    for d in data.aws_ami.nix_tailscale.block_device_mappings : d
+    if d.device_name == local.state_device_name
+  ][0]
+}
+
 resource "aws_instance" "factorio" {
-  ami                    = "ami-07df5833f04703a2a"
-  instance_type          = "t3.micro"
-  key_name               = aws_key_pair.deploy_key.key_name
-  availability_zone      = local.aws_az
-  iam_instance_profile   = aws_iam_instance_profile.factorio.name
-  vpc_security_group_ids = [aws_security_group.factorio.id]
+  ami                         = data.aws_ami.nix_tailscale.id
+  instance_type               = "t3.micro"
+  key_name                    = aws_key_pair.deploy_key.key_name
+  associate_public_ip_address = false
+  availability_zone           = local.aws_az
+  iam_instance_profile        = aws_iam_instance_profile.factorio.name
   # Do not create until the ssm parameter has been created
   depends_on = [aws_ssm_parameter.factorio_tailnet_key]
+
   root_block_device {
-    volume_size = 8
+    volume_type = "gp3"
   }
 
-  user_data = <<EOT
-#!/usr/bin/env bash
-nix-shell -p bash util-linux e2fsprogs --run bash <<'EOF'
-  set -e
-
-  function wait_for() {
-    drive="$1"
-    until [ -e "$drive" ]; do
-      echo "Waiting for $drive to be ready..."
-      sleep 1
-    done
-    echo "Device $drive ready"
+  ephemeral_block_device {
+    device_name = local.state_device_name
+    no_device   = true
   }
-
-  fallocate -l 1G /swapfile
-  chmod 600 /swapfile
-  mkswap /swapfile
-  swapon /swapfile
-
-  drive="/dev/sdf"
-  wait_for "$drive"
-  if [ "ext4" != "$(blkid -o value -s TYPE "$drive")" ]; then
-    mkfs -t ext4 -L state "$drive"
-  fi
-
-  drive="/dev/disk/by-label/state"
-  wait_for "$drive"
-  mkdir /state
-  mount "$drive" /state
-EOF
-EOT
 }
 
+# Manage the state device separately from the instance AMI to avoid
+# recreation if the instance is recreated.
+# TODO: Test this better
+# TODO: Importing an existing disk
 resource "aws_ebs_volume" "factorio" {
   availability_zone = local.aws_az
-  size              = 2
+  snapshot_id       = local.ami_device.ebs.snapshot_id
+  type              = local.ami_device.ebs.volume_type
 }
 
 resource "aws_volume_attachment" "factorio" {
-  device_name = "/dev/sdf"
+  device_name = local.state_device_name
   volume_id   = aws_ebs_volume.factorio.id
   instance_id = aws_instance.factorio.id
 }
 
-output "factorio_ip_addr" {
-  value = aws_instance.factorio.public_ip
+# Ensure that the instance is reachable via `ssh` before deploying
+resource "null_resource" "wait" {
+  provisioner "remote-exec" {
+    connection {
+      user = "root"
+      host = "factorio"
+    }
+
+    inline = [":"]
+  }
+}
+
+local {
+  ssh_opts = "-o StrictHostKeyChecking=accept-new"
+}
+
+resource "null_resource" "deploy" {
+  provisioner "local-exec" {
+    # interpreter = "nix develop ${path.module}# --command bash"
+    command = "deploy --ssh-opts=\"${ssh_opts}\" ${path.module}#factorio"
```

Blocking: Locals must be prefixed with `local.` in order to build.
```suggestion
    command = "deploy --ssh-opts=\"${local.ssh_opts}\" ${path.module}#factorio"
```

---

Don't commit late at night kids...